Scoped `impl Trait for Type` (#3634)
rust-lang/rfcs: text/3634-scoped-impl-trait-for-type.md

Global implementations can be imported from the root namespace, for example to shadow a scoped implementation:

```rust
use ::{impl Trait for Type};
```

### Scoped implementations and generics
[scoped-implementations-and-generics]: #scoped-implementations-and-generics

Scoped implementations are resolved on most generics' type arguments where those are specified, and become part of the (now less generic) host type's identity:

```rust
#[derive(Default)]
struct Type<T>(T);

trait Trait {
    fn trait_fn();
}

impl<T: Trait> Type<T> {
    fn type_fn() {
        T::trait_fn();
    }
}

mod nested {
    use super::{Trait, Type};

    use impl Trait for () {
        fn trait_fn() {
            println!("nested");
        }
    }

    pub type Alias = Type<()>;
}
use nested::Alias;

fn main() {
    Alias::type_fn(); // "nested"

    // Type::<()>::type_fn();
    //             ^^^^^^^ error[E0599]: the function or associated item `type_fn` exists for struct `Type<()>`, but its trait bounds were not satisfied

    // let t: Type<()> = Alias::default();
    //                   ^^^^^^^^^ error[E0308]: mismatched types

>     let t: Type<() as Trait in nested> = Alias::default();

Author: Yes, this is a mistake.  An inline implementation environment can only refer to implementations that it can actually see.

Fixed in 129e734ae1bc198becb49b2b4590232e0a746c5e.

[rustdoc-documentation-changes]: #rustdoc-documentation-changes

### `use` and `impl` keywords

The documentation pages [for the `use` keyword] and [for the `impl` keyword] are adjusted to (very) briefly demonstrate the respective scoped use of `impl Trait for Type`.

[for the `use` keyword]: https://doc.rust-lang.org/stable/std/keyword.use.html
[for the `impl` keyword]: https://doc.rust-lang.org/stable/std/keyword.impl.html

### `TypeId`

The page for [`TypeId`] gains two sections with the following information:

```markdown
# `TypeId` and scoped implementations

To make sure that that are no mix-ups between, for example, `HashSet<T>` and
`HashSet<T as Hash in module>`, any such difference implies distinct `TypeId`s between
such discretised generics (and that the types are not mutually assignable).

This also affects trait-bounded generic type parameters: If `T` is bounded on `Hash`, then
`TypeId::of::<T>()` results in distinct `TypeId`s in that context depending on the
captured implementation.

However, note that `TypeId::of::<T>()` and `TypeId::of::<T as Hash in module>()` are
always equivalent for one definition of `T`, as `TypeId::of`'s implementation does **not**

> have a `T: Hash` bound!

Author: `Type as Trait in module` is technically not a type at all, as implementations never "stick" to the outside of a type.
It's a type argument consisting of a type and an inline implementation environment.

This combination, along with any scoped implementations inherited from the surrounding scope, is "baked" into the effective (opaque) type of the generic type parameter as part of monomorphisation.
These opaque types are special-cased in terms of how the baked implementation environment influences the observed `TypeId`:

- If the type parameter is wrapped in an implementation-aware generic (even indirectly), e.g. as `TypeId::of::<HashSet<T>>()`, then the `TypeId` identifies the combination of the discrete type given as part of the type argument and the *full* set of baked implementations.
  (Normally this would happen because the e.g. `HashSet<T>` is used as type argument and *then* has its `TypeId` taken inside another layer of generic function call.)

  This is necessary to ensure existing `unsafe` code doesn't accidentally transmute hashtables in violation of their coherence expectations, i.e. part of version 1 of "the hashtable problem".

  (Writing it out explicitly like above, without indirection through another generic call, also produces the [Warning: `TypeId` of implementation-aware generic discretised using generic type parameters](https://github.com/Tamschi/rust-rfcs/blob/scoped_impl_trait_for_type/text/3634-scoped-impl-trait-for-type.md#behaviour-changewarning-typeid-of-implementation-aware-generic-discretised-using-generic-type-parameters), since this is almost never what you actually want to observe when making type-keyed collections.)

- In all other cases, the observed `TypeId` identifies the combination of the type given as type argument and any bounds-relevant captured implementations.

  This makes type-keyed and type-erased collections work as expected by consumer code, without over-distinguishing entries based on scoped implementations that are irrelevant to their function (which is what I call 'version 2 of "the hashtable problem"').

  You can find more on the reasoning here in the section on [Logical consistency of type-erased collections](https://github.com/Tamschi/rust-rfcs/blob/scoped_impl_trait_for_type/text/3634-scoped-impl-trait-for-type.md#of-type-erased-collections).

Please let me know if this works as clarification.
I should add a cross-reference to the [Logical consistency](https://github.com/Tamschi/rust-rfcs/blob/scoped_impl_trait_for_type/text/3634-scoped-impl-trait-for-type.md#logical-consistency) explanations here in any case, though.

> have a `T: Hash` bound!

Reviewer: Here is my understanding:
- When a concrete type is used as a type argument, the observed type-identity of the corresponding type parameter in the generic item *changes* depending on the where-bounds present on the type parameter.

This seems strange, to say the least. As an example:
```rust
trait TypeContainer {
  type Type;
}

struct Generic<T>(T) /* where T: Marker */;
impl<T> TypeContainer for Generic<T> {
  type Type= T;
}

trait Marker {}

mod one {
  use impl super::Marker for () {}
  pub type Alias = super::Generic<()>;
}

mod two {
  use impl super::Marker for () {}
  pub type Alias = super::Generic<()>;
}

type X = ();
type Y = <one::Alias as TypeContainer>::Type;
type Z = <two::Alias as TypeContainer>::Type;

// or, equivalently
// type Y = <Generic<() as Marker in one> as TypeContainer>::Type;
// type Z = <Generic<() as Marker in two> as TypeContainer>::Type;
```

Is it then the case that X, Y, and Z are the same if there is no `T: Marker` bound, but distinct if there is one (despite all three types being known to be `()`)?

> have a `T: Hash` bound!

Author: > Or is it the case that; no, all three types are known to be identical _in concrete code_, but _generic_ code defining these three types are not only not known to be identical, but are in fact _known to be distinct_ (as witnessed by runtime type comparison).
> 
> […]

This was my intention, but I failed to consider that generic type parameters can flow back into concrete code transparently like this. I'll have to go over this (and the issue) carefully and see how my ideas hold up and where I need to specify more thoroughly or change something.

(_Tentatively_, I think that the implementations can/should 'fall off' in concrete code so that `X`, `Y`, `Z` and `()` are known to be identical and would yield identical `TypeIds`, but that this would not regress the issue because the different concrete `Generic<…>`s are not at all assignable to each other (regardless of whether the bound is present or not).)

I can't look into this properly quite right now unfortunately, but hopefully I'll manage to do it at some point this week.

> have a `T: Hash` bound!

Reviewer: > Tentatively, I think that the implementations can/should 'fall off' in concrete code so that X, Y, Z and () are statically known to be identical and would/could¹ yield identical TypeIds

In effect, implicitly making the type of a generic type parameters a sort of `Wrapper<T>` with partial implementation-awareness.

On a related note, I wonder if you can automate implementation-awareness based on `impl` blocks (e.g. `HashMap` relies on `K: Eq + Hash` in one of its impl blocks, so functions from that impl block carry implementation-awareness, rather than the type itself?

> have a `T: Hash` bound!

Author: > > Tentatively, I think that the implementations can/should 'fall off' in concrete code so that X, Y, Z and () are statically known to be identical and would/could¹ yield identical TypeIds
> 
> In effect, implicitly making the type of a generic type parameters a sort of `Wrapper<T>` with partial implementation-awareness.

In a way. I'll think about whether I explain it that way to some extent, though it's possible it's too not-quite-the-same (for me) to fit it into the text well.

> On a related note, I wonder if you can automate implementation-awareness based on `impl` blocks (e.g. `HashMap` relies on `K: Eq + Hash` in one of its impl blocks, so functions from that impl block carry implementation-awareness, rather than the type itself?

Implementation-awareness must be shared between distinct impl blocks.

For example, the `std`'s `HashMap` has inherent, `Extend`, `From`, `FromIterator`, `Index`, `PartialEq` and `Eq` `impl`s that expect `K: Hash` to be coherent. The category likely also should include external implementations in at least some cases, which for simplicity would mean all cases.

Having that information "flow backwards" from `impl`s to overall `type`s' identities then would introduce a lot of tricky behaviour impacts at a distance, and most likely wouldn't be compatible with early code generation for concrete code.

## No external scoped implementations of sealed traits
[no-external-scoped-implementations-of-sealed-traits]: #no-external-scoped-implementations-of-sealed-traits

Consider this library crate:

```rust
pub struct Generic<T>(T);

mod private {
    // Implemented only on traits that are also `Sealed`.
    pub trait Sealing {}
}
use private::Sealing;

pub trait Sealed: Sealing {
    fn assumed {
        // ❷
    }
}

impl<T: Sealed> Generic {
    fn assuming {
        // ❶
    }
}
```

In this crate, any code at ❶ is currently allowed to make safety-critical assumptions about code at ❷ and other implementations of `assumed`.

To ensure this stays sound, scoped `impl Trait for Type` where `Trait` is external requires that all supertraits of `Trait` are visible to the crate defining the scoped implementation or are defined not in `Trait`'s definition crate (meaning they must still be exported from a crate somewhere in the dependency tree).

See also [scoped-implementation-of-external-sealed-trait].

> ## Type parameters capture their *implementation environment*

Author: I just edited this section (and some related wording that I caught) a bit to effectively use this terminology in ef35245c61d5f9d459cd93593a7bebead2143221 ¹, but adding a section that defines the terms formally is a great idea.

I used 'concrete' and (mainly) 'discrete' interchangeably in this RFC so far, but it's also a good idea to be consistent about that. (I personally would tend to use 'discrete' for types and 'concrete' for values, but that's purely based on vibes so I'll do a bit of research tomorrow to see if there's any precedent in that regard.)

(I should be able to add the definition properly in the next one or two days, depending on how busy I am.)

¹ I'm not sure it was you who pointed this out on Discord. If not: Thank you for the review! If yes: Thanks again!

\*\* but see [which-structs-should-be-implementation-invariant].

See also [why-specific-implementation-invariant-generics].

[call expressions]: https://doc.rust-lang.org/reference/expressions/call-expr.html#call-expressions

## Call expressions' *function operand* captures its *implementation environment*
[call-expressions-function-operand-captures-its-implementation-environment]: #call-expressions-function-operand-captures-its-implementation-environment

Call expressions capture the *implementation environment* in their *function operand*, acting as host for [implementation-invariant-generics]. This enables call expressions such as

```rust
Option::<Type as Debug in module>::fmt(…)
```

where `fmt` receives the specified scoped implementation by observing it through the `T: Debug` bound on its implementing `impl` block.

If no observing bound exists, code of this form should produce a warning spanning the `Trait in module` tokens. (see [unused-scoped-implementation])

## Type aliases are opaque to scoped implementations
[type-aliases-are-opaque-to-scoped-implementations]: #type-aliases-are-opaque-to-scoped-implementations

> As scoped `impl Trait for Type` is a fully lexically-scoped feature, the *implementation environment* present in a scope does not affect types hidden behind a type alias, except for the top-level type directly:

Author: No, as that is an implementation on the 'top-level' type `Type`.

The opacity means that the implementation environments captured by any type arguments in the type alias's definition (i.e. those applicable to `Type` in `pub type Alias = Generic<Type>;`) are frozen and don't ever vary based on the type alias's usage elsewhere.

Uses of `Alias` are still affected by the implementation environment for the concrete type `Generic<Type>`, including blanket implementations like `use impl<T> Trait for Generic<T> { … }` (if `Type: Sized`).

If the alias contains type variables, e.g. `type Alias<T> = Generic<T>;`, then the implementation environment of the type parameter `T` is captured as part of `Generic<T>` in this definition and frozen, but the implementation environment of type arguments given to `Alias<T>` in also inserted "below" that as the type is discretised further.

(I *think* this is worded okay in the RFC here, but maybe it's not verbose enough. I'll mull it over.)

```rust
trait Trait {
    fn method(&self) -> &str;
}

impl Trait for Type {
    fn method(&self) -> &str {
        "global"
    }
}

mod m1 {
    use super::Type;

    pub type Alias = [Type; 1];
}

mod m2 {
    use super::{Type, Trait};

    pub use impl Trait for Type {
        fn method(&self) -> &str {
            "scoped"
        }
    }

    pub use impl<T: Trait> Trait for [T; 1] {
        fn method(&self) -> &str {
            self[0].method()
        }
    }
}

fn main() {
    use m1::Alias;
    use m2::{
        impl Trait for Type,
        impl Trait for [Type; 1],
    };

    assert_eq!([Type].method(), "scoped");
    assert_eq!(Alias::default().method(), "global");
}
```

Scoped implementations may still be observed through bounded generic type parameters on the type alias itself. (see [binding-choice-by-implementations-bounds])

> ## `TypeId` of generic type parameters' opaque types
> [typeid-of-generic-type-parameters-opaque-types]: #typeid-of-generic-type-parameters-opaque-types
> 
> In addition to the type identity of the specified type, the `TypeId` of opaque generic type parameter types varies according to the captured *implementation environment*, but *only according to implementations that are relevant to their bounds (including implicit bounds)*, so that the following program runs without panic:
> 

Author: I could word this better. The type of generic type parameters is always opaque, as it abstracts the various concrete types inserted in its place during monomorphisation (but also, due to implementation environment "baking", these concrete types can be effectively inexpressible in Rust).

```suggestion
## `TypeId` of generic type parameters' opaque types
[typeid-of-generic-type-parameters-opaque-types]: #typeid-of-generic-type-parameters-opaque-types

In addition to the type identity of the specified type, the `TypeId` of generic type parameter types varies according to the *implementation environment* that has been captured into them, but *only according to implementations that are relevant to their bounds (including implicit bounds)*, so that the following program runs without panic:
```

> 

Author: Resolved in f3589adf495ceb828627817e234f4dffdaedc5b8, hopefully, but feel free to unresolve if it's still unclear.

```rust
use std::any::TypeId;

#[derive(Default)]
struct Type;
trait Trait {}
impl Trait for Type {}

#[derive(Default)]
struct Generic<T>(T);

mod nested {
    pub(super) use impl super::Trait for super::Type {}
}

// `A` and `B` are distinct due to different captured implementation environments.
type A = Generic<Type>;
type B = Generic<Type as Trait in nested>;

> fn no_bound<T: 'static, U: 'static>(_: Generic<T>, _: Generic<U>) {
>     assert_eq!(TypeId::of::<T>(), TypeId::of::<U>());
>     assert_ne!(TypeId::of::<Generic<T>>(), TypeId::of::<Generic<U>>());
> 
>     assert_eq!(TypeId::of::<T>(), TypeId::of::<Type>());
>     assert_eq!(TypeId::of::<U>(), TypeId::of::<Type>());
> }

Author: This is the same `TypeId` special-casing when type parameters are wrapped in implementation-aware generics that I mentioned in [this comment](https://github.com/rust-lang/rfcs/pull/3634/files/b7fb5e2183eb4a0f5aa21fb7835c85965f7ca9d8#r1599053101), but I should clarify that the respective lines produce warnings.

```suggestion
fn no_bound<T: 'static, U: 'static>(_: Generic<T>, _: Generic<U>) {
    assert_eq!(TypeId::of::<T>(), TypeId::of::<U>());
    assert_ne!(TypeId::of::<Generic<T>>(), TypeId::of::<Generic<U>>()); // ⚠⚠

    assert_eq!(TypeId::of::<T>(), TypeId::of::<Type>());
    assert_eq!(TypeId::of::<U>(), TypeId::of::<Type>());
}
```

fn yes_bound<T: Trait + 'static, U: Trait + 'static>(_: Generic<T>, _: Generic<U>) {
    assert_ne!(TypeId::of::<T>(), TypeId::of::<U>());

>     assert_ne!(TypeId::of::<Generic<T>>(), TypeId::of::<Generic<U>>());

Author: ```suggestion
    assert_ne!(TypeId::of::<Generic<T>>(), TypeId::of::<Generic<U>>()); // ⚠⚠
```

(The code block is messing with the suggestion, so I have to split this to do the edit here.)

    assert_eq!(TypeId::of::<T>(), TypeId::of::<Type>());
    assert_ne!(TypeId::of::<U>(), TypeId::of::<Type>());
}

fn main() {
    no_bound(A::default(), B::default());
    yes_bound(A::default(), B::default());
}
```

> 

Author: ```suggestion

(The lines marked with ` // ⚠⚠` produce warnings: [behaviour-changewarning-typeid-of-implementation-aware-generic-discretised-using-generic-type-parameters])

```

This is intentional, as it makes the following code trivial to reason about:

```rust
{
    use a::{impl TheTrait for TheType}; // <-- Clearly unused, no hidden interdependencies.
    {
        use b::{impl TheTrait for TheType};
        // ...
    }
}
```

(The main importance here is to not allow non-obvious dependencies of imports. Implementations can still access associated items of a *specific* other implementation by bringing it into a nested scope or binding to its associated items elsewhere. See also [independent-trait-implementations-on-discrete-types-may-still-call-shadowed-implementations].)

## Binding choice by implementations' bounds
[binding-choice-by-implementations-bounds]: #binding-choice-by-implementations-bounds

Implementations bind to other implementations as follows:

| `where`-clause¹ on `impl`? | binding-site of used trait | monomorphised by used trait? |
|-|-|-|
| Yes. | Bound at each binding-site of `impl`. | Yes, like-with or as-part-of type parameter distinction.  |

> | No. | Bound once at definition-site of `impl`. | No. |

Author: That's definitely a much better way to word this than mine, yes! Thank you.

More precisely, bounds on generic type parameters are satisfied by generic type arguments (and accordingly the relevant bindings are sourced from there) while all other bindings are *not* influenced by generic type arguments whatsoever.
This makes the scopes strictly lexical and keeps monomorphisation of functions in check at the same level as what you would expect when using newtypes.

I'll revise this section, though I'm not sure yet that I can also simplify the example.

> It may be better to allow `use impl FnBoundedMonomorphic for Type {}` without `where_monomorphic_subtrait` in ❷ and disallow incompatible unsizing instead. I'm not sure about the best approach here.

In ❸, `FnBoundedMonomorphic` is monomorphically reimplemented for `Type`, which means it "picks up" the scoped `Type: Trait` implementation that's in scope there from ❷.

### `NestedMonomorphic` / `::nested_monomorphic`

The global implementation of `NestedMonomorphic` in ❸ the binds to the scoped implementation of `Trait` on `Type` from ❷ internally. This allows outside code to call into that function indirectly without exposing the scoped implementation itself.

### `BoundedOnOther` / `::bounded_on_other`

As this discrete implementation's bound isn't over the `Self` type (and does not involved generics), it continues to act only as assertion and remains monomorphic.

## Binding and generics

`where`-clauses without generics or `Self` type, like `where (): Debug`, **do not** affect binding of implementations within an `impl` or `fn`, as the non-type-parameter-type `()` is unable to receive an *implementation environment* from the discretisation site.

> 

Author: Yes, exactly that. (More precisely, such a scoped implementation *at the caller* does not affect the behaviour of the function, but *at its definition* it does, and the trivial bound is satisfied according to the function definition's implementation environment too.)

I think that, in this case, it's overall better to keep monomorphisation in check and not turn discrete-as-written implementations into monomorphised ones. Rustdoc *does* display trivial bounds though 🤔
(Another practical issue is inability to write implementation environments inline for anything that's not a generic type argument… but that could just be a case of me personally coming up blank if I think about the syntax for it.)

Changing this doesn't sit well with me because it seems unintuitive (it would make `where`-clauses a form of first-class implicit-only parameter) and there may be macros that emit spurious trivial bounds. It would also break that trivial bounds can be used as assertions, at least if allowed without global implementation… though the `static_assertions` crate notably does not currently do this.

In general, I'm very hesitant to add anything that reinterprets currently valid Rust code, even if I'm pretty sure this wouldn't be a compilation-breaking change. Maybe it would be a better idea to hide trivial bounds from the docs instead, going for consistency in the other direction?

> 

Reviewer: Makes sense to me, but you might consider explicitly calling out that trivial bounds are not affected by scoped implementations.

## Shadowing with different bounds

Scoped implementations may have different bounds compared to an implementation they (partially) shadow. The compiler will attempt to satisfy those bounds, but if they are not satisfied, then the other implementation is not shadowed for that set of generic type parameters and no additional warning or error is raised.

(Warnings for e.g. unused scoped implementations and scoped implementations that only shadow a covering global implementation are still applied as normal. It's just that partial shadowing with different bounds is likely a common use-case in macros.)

```rust
struct Type1;
struct Type2;

trait Trait1 {
    fn trait1() {
        println!("1");
    }
}
impl<T> Trait1 for T {} // <--

trait Trait2 {
    fn trait2() {
        println!("2");
    }
}
impl Trait2 for Type2 {} // <--

trait Say {
    fn say();
}
impl<T: Trait1> Say for T
where
    T: Trait1, // <--
{
    fn say() {
        T::trait1();
    }
}

{
    use impl<T> Say for T
    where
        T: Trait2 // <--
    {
        fn say() {
            T::trait2();
        }
    }

    Type1::say(); // 1
    Type2::say(); // 2
}
```

> ## No priority over type-associated methods
> 
> Scoped `impl Trait for Type` has *the same* method resolution priority as an equivalent global implementation would have if it was visible for method-binding in that scope. This means that directly type-associated functions still bind with higher priority than those available through scoped implementations.

Author: Yes, I wasn't aware that's the term for it. (For reference, [here's the Reference](https://doc.rust-lang.org/reference/items/implementations.html).)

```suggestion
## Resolution priority

Scoped `impl Trait for Type` has *the same* resolution priority as an equivalent global trait implementation would have if it was visible in that scope. This means that there may be ambiguity between scoped implementations introduced at different nesting levels, just as between equivalent global trait implementations.

> Flagging a potentially unintended behaviour change with this error seems like the safer choice, and it's easy to resolve using a qualified path where needed.

Inherent implementations still always shadow all trait implementations, including scoped ones.
```

[`DynMetadata<dyn Display>`]: https://doc.rust-lang.org/stable/core/ptr/struct.DynMetadata.html

## Interaction with return-position `impl Trait`

Consider the following functions:

```rust
trait Trait {}

fn function() -> impl Trait {
    use impl Trait for () {}

    () // Binds on trailing `()`-expression.
}

fn function2() -> impl Trait {
    use impl Trait for () {}

    {} // Binds on trailing `{}`-block used as expression.
}
```

In this case, the returned opaque types use the respective inner scoped implementation, as it binds on the `()` expression.

The following functions do not compile, as the implicitly returned `()` is not stated *inside* the scope where the implementation is available:

```rust
trait Trait {}

fn function() -> impl Trait {
                 ^^^^^^^^^^
    use impl Trait for () {}
    ---------------------

    // Cannot bind on implicit `()` returned by function body without trailing *Expression*.
}

fn function2() -> impl Trait {
                  ^^^^^^^^^^
    use impl Trait for () {}
    ---------------------

    return; // Cannot bind on `return` without expression.
    -------
}

> ```

Author: In theory it could be made to, but I (subectively) think that would be hard to follow, since for example other items nested in function bodies aren't available to their signature (if I'm not completely mistaken).

The function body block isn't visually located inside itself, so to me it appear one or a half level of scoping higher than a scoped `impl Trait for Type` placed inside it.

> ```

Reviewer: This seems like too subtle of a distinction to make. I don't think most rust programmers conceptualize the implicit return value to be any different than an explicit return value. While, yes, there are already subtle differences with regards to implicit drop and lifetimes, I don't think this RFC should add to those pain points by distinguishing between implicit return, return with expression, and return without expression.

> ```

Author: I don't have a strong preference in this regard. `return` without expression at least could definitely be defined to pick up the environment on the `return` token instead without it feeling odd in the long run, so I'll change that here when I can.

If I had totally free reign, I would probably put off the matter with the fully implicit return until the very last minute and then run a survey, but I don't know if that's *a thing* 😅
It's really hard for me to tell whether reaching the end of the function body should observe the implementation environment just outside or just inside the function, though. From a purely practical/convenience point of view, I think I would say 'inside' too, and I now suspect it may also be closer to developer intent in mixed scenarios, so I'll change it to that.

(Sorry, today I'm having some trouble with the wider reason it took me so long to submit this in the first place, so I'll be a bit slow to implement changes and respond to the larger points you brought up.)

        ()
    }
}
```

This compiles because the end of the function is not reachable:

```rust
trait Trait {}

fn function() -> impl Trait {
    {
        use impl Trait for () {}
        return (); // Explicit `return` is required to bind in the inner scope.
    }
}
```

## Static interception of dynamic calls

As a consequence of binding outside of generic contexts, it *is* possible to statically wrap *specific* trait implementations on *concrete* types. This includes the inherent implementations on trait objects:

```rust
use std::fmt::{self, Display, Formatter};

{
    use impl Display for dyn Display {
        fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
            // Restore binding to inherent global implementation within this function.
            use ::{impl Display for dyn Display};

            write!(f, "Hello! ")?;
            d.fmt(f)?;
            write!(f, " See you!")
        }
    }

    let question = "What's up?"; // &str
    println!("{question}"); // "What's up?"

    let question: &dyn Display = &question;
    println!("{question}"); // Binds to the scoped implementation; "Hello! What's up? See you!"
}

> ```

Author: This has to be explicitly special-cased to make it an error (also for supertraits, now that we have dyn-upcasting), but you're right. Even if someone later figures out a **really** good reason to have this, it's safer to start without it.

What about `&dyn Trait` types? Calling a function that takes `impl Trait` with a `&dyn Trait` seems like a common pattern for (mut-)ref-transitive traits like `Display` or `Write`, but this transitivity is never a built-in to my knowledge.

> ```

Author: Hm… this is actually pretty complicated.

Would this also exclude type aliases like `type Alias = dyn Trait;`? What about `type Alias = impl Trait;`?

That aside:

- Making it a hard error in all cases could interfere with blanket implementations like `use impl<T: ?Sized> Trait for T { … }` and equivalent imports.

- Making it silently ineffective for built-in trait implementations of trait objects would introduce subtle hole-punching:

  ```rust
  use a::Trait;
  use b::{impl<T: ?Sized> Trait for T};

  #[derive(Trait)] // Recursive-into-fields implementation.
  struct Struct {
      x: X,
      y: Y,
      z: dyn Trait,
  }
  ```

  Is the scoped implementation effective for recursion into `z` here? It would depend on implementation details of the derive-macro, as for example functions like

  ```rust
  pub Trait {
      fn by_ref<T: Trait>(reference: &T);
      fn by_value(value: impl Trait);
  }
  ```

  or alternatively functions like

  ```rust
  pub Trait {
      fn access(&self);
      fn consume(self) where Self: Sized;
  }
  ```

  would each bind differently between them when given a reference.

  (I am *assuming* that the derive-macro properly uses fully-qualified calls. If it uses method-style calls like `self.z.access()` and `self.z.consume()`, this might even differ by whether there's a `+ Copy` on the field type, though I'm not sure about that.)

Maybe the best course of action here would be to make it a **deny-by-default lint** that like other warnings is silenced if its span is resolved other-than-call-site? That would make everything behave consistently where there's an encapsulation boundary from an external macro, alert the user of the issue, but also provide an "I know what I am doing; this is intentional"-button in form of `#[allow(…)]`.

(Or at least _should_ behave like that for macros in an ideal world. In practice most proc macros are unhygienic in my experience, which is another case where a way to `#[allow(…)]` it may be needed.)

> ```

Reviewer: The blanket impl `impl<T: ?Sized> Trait for T` already excludes `dyn Trait`, so there is no issue there. I would expect the equivalent `use impl` to also exclude `dyn Trait`. I would also prohibit scoped trait implementations of `Trait` on aliases for `dyn Trait`, and prohibit scoped implementations of `Trait` on type-alias `impl Trait`. I'm not as sure about scoped implementations of _other_ traits on type-alias `impl Trait`.

> ```

Author: Fixed in 99dc57251c6e03b8f333d1acf551fb6c0ab6936b.

(Thanks for explaining this to me.)

This doesn't matter for `Span::call_site()` vs. `Span::mixed_site()` since scoped implementations would resolve transparently through both, but it does matter for `Span::def_site()` which should exclude them.

It very much does matter if one of the opt-in mitigations for [first-party-implementation-assumptions-in-macros] is implemented.

## Which `struct`s should be implementation-invariant?
[which-structs-should-be-implementation-invariant]: #which-structs-should-be-implementation-invariant

This is a tough question because, runtime behaviour difference of [of-type-erased-collections] aside, the following makes shifting a type from [implementation-aware-generics] to [implementation-invariant-generics] a compilation-breaking change:

```rust
struct Type;
struct Generic<T>(T);
trait Trait {}

mod a {
    use super::{Type, Generic, Trait};
    pub use impl Trait for Type {}
    pub type Alias = Generic<T>;
}

mod b {
    use super::{Type, Generic, Trait};
    pub use impl Trait for Type {}
    pub type Alias = Generic<T>;
}

use impl Trait for a::Alias {}
use impl Trait for b::Alias {}

> ```

Reviewer: Do we want to be able to have distinct `impl Trait for Generic<Type as X in one>` and `impl Trait for Generic<Type as X in two>`? This seems unexpected.

Author: It's required for backwards-compatibility, or at least having a distinct `impl Trait for Generic<Type>` that isn't available for `Generic<Type as X in one>` or `Generic<Type as X in two>` is. Existing code can make safety-critical assumptions about the exact behaviour of the former.

(Forbidding implementations for concrete types that contain captured scoped implementations outright is technically also possible, but I don't think that would helpful.)

Fitting these distinct implementation targets into Rustdoc could be interesting, though, just in terms of information arrangement.

## Glue crate suggestions
[glue-crate-suggestions]: #glue-crate-suggestions

If crates move some of their overlay features into glue crates, as explained in [unblock-ecosystem-evolution], it would be nice if they could suggest them if both they and e.g. Serde were `cargo add`ed as direct dependencies of a crate currently being worked on.

An example of what this could look like:

```toml
[package]
name = "my-crate"
version = "0.1.2"
edition = "2021"

[dependencies]
# none

[suggest-with.serde."1"]
my-crate_serde_glue = "0.1.0"

[suggest-with.bevy_reflect."0.11"]
my-crate_bevy_reflect_glue = "0.1.2"

[suggest-with.bevy_reflect."0.12"]
my-crate_bevy_reflect_glue = "0.2.1"
```

(This sketch doesn't take additional registries into account.)

Ideally, crates.io should only accept existing crates here (but with non-existing version numbers) and Cargo should by default validate compatibility where possible during `cargo publish`.

## Reusable limited-access APIs

Given a newtype of an unsized type, like

```rust
#[repr(transparent)]
pub struct MyStr(str);
```

> for example, there is currently no safe-Rust way to convert between `&str` and `&MyStr` or `Box<MyStr>` and `Box<str>`, even though *in the current module which can see the field* this is guaranteed to be a sound operation.

Reviewer 2: We need something like this for the delegation proposal as well
https://github.com/rust-lang/rfcs/blob/923b8e65639de80a93ef2b9770ba1c976c24cf5d/text/0000-fn-delegation.md#newtype-support-1
https://github.com/rust-lang/rfcs/blob/923b8e65639de80a93ef2b9770ba1c976c24cf5d/text/0000-fn-delegation.md#newtype-support

Interfaces for going from
- `&Newtype<T>` to `&T`
- `&mut Newtype<T>` to `&mut T`
- `Newtype<T>` to `T`
- `T` to `Newtype<T>`

for `struct Newtype<T>(pub(vis) T)`, *but* not usable outside of `vis` (the wrapped field's visibility area).

For structures that have a single field we can just hardcode such logic into the compiler, but it would be nice to be a bit more extensible.
